### PR TITLE
Javascript javascript1 week4 js

### PR DESCRIPTION
**Pull request by tanya-newStart**
Some tasks that use arrays and objects.
Voice_assistant
Find_danishLetters in a string

### CODE REVIEW

File: javascript/javascript1/week4/voice_assistant.js [hunk 1]
```diff
@@ -1,68 +1,81 @@
...
+  [patterns.addPattern]: {
+    reply: (input) => {
+      return addToDo(input);
+    },
+  },
+  [patterns.removePattern]: {
+    reply: (input) => {
+      return removeToDo(input);
+    },
+  },
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/voice_assistant.js:38]: Very minor syntactic adjustment here but be as consistent. Below you write the functions without the curly brackets and return  { return } and here with. There's arguments to have have either but consistency is key.

File: javascript/javascript1/week4/voice_assistant.js [hunk 2]
```diff
@@ -136,22 +156,58 @@ function calculator(a, b, operation) {
...
+    ];
+    for (let i = 0; i < clothes.length; i++) {
+      const tempRange = clothes[i][1];
+      if (temp >= tempRange[0] && temp < tempRange[1]) {
+        return clothes[i][0];
+      }
+    }
+    return `Stay home! It's freezing!`;
+  }
+}
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/voice_assistant.js:200]: Nice refactor! Though it seems it crept into the voice_assistant.js commit. It looks like you're having fun with the math and logic which is great to see!

File: javascript/javascript1/week4/voice_assistant.js [hunk 3]
```diff
@@ -136,22 +156,58 @@ function calculator(a, b, operation) {
...
-//console.log(greetings("Hello, my name is Tanya"));
-console.log(addToDo("Add dishes to my to do list."));
+
+function setTimer(input) {
+  const match = input.match(patterns.timerPattern);
+  if (match && match.length === 2) {
+    const minutes = parseInt(match[1]);
+    setTimeout(function () {
+      console.log("Time is up!");
+    }, minutes * 60 * 1000);
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/voice_assistant.js:176]: Great and works as expected. You have the fundamentals down pat but in the future of the your Javascript journey you will notice how setting timeouts can give you wonky behaviour (especially in React) then multiple setTimeouts will run on refresh.

Here you will become familiar with the importance clearing your setTimeouts (clearTimeout). It's definitely beyond what we have learnt but you are far ahead and may want some extra reading then read this [article on setTimout & clearTimeout](https://alonewolf24.medium.com/mastering-javascripts-settimeout-and-cleartimeout-functions-7eb63a369175)

File: javascript/javascript1/week4/voice_assistant.js [hunk 4]
```diff
@@ -1,68 +1,81 @@
...
-    return `You have ${toDoList.length} items in your to do list: ${toDoList}`;
-  } else if (input === "What day is it today?") {
-    return currentDate();
-  } else if (mathPattern.test(input)) {
-    return calculateResult(input);
+  for (const pattern in patterns) {
+    if (patterns[pattern].test(input)) {
+      const replyFunction = commands[patterns[pattern]].reply;
+      return replyFunction(input);
+    }
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/voice_assistant.js:72]: Very minimal and sleek solution which you can easily expand upon with new rules 👏

**Review by islanderupnorth — COMMENTED** [review]
Really great work Tanya!

**Review by islanderupnorth — COMMENTED** [review]
Great work again. If you want an extra challenge perhaps attempt to re-write this function but using the .reduce array method?

[This](https://dev.to/mattlewandowski93/arrayreduce-is-goated-1f1j?ref=dailydev) popped up on my feed today and maybe you will find interesting as well. Feel free to skip this if you want to but dropping in case you want to give it a try!

File: javascript/javascript1/week4/voice_assistant.js (hunk 4 above)
Comment (tanya-newStart) [reply javascript/javascript1/week4/voice_assistant.js:72]: Thank you!

File: javascript/javascript1/week4/playground/credit_card_type.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  { dankort: /^(5019)\d+$/ },
+];
+function getCardInfo(num) {
+  for (const card of creditTypes) {
+    for (const [type, regex] of Object.entries(card)) {
+      if (regex.test(num)) {
+        return type;
+      }
+    }
+  }
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/playground/credit_card_type.js:22]: You could simplify this if you swap the looping to loop `creditTypes` instead to the following:

```
for (const [type, regex] of Object.entries(creditTypes)) {
    if (regex.test(num)) {
      return type;
    }
  }
```

File: javascript/javascript1/week4/voice_assistant.js (hunk 3 above)
Comment (tanya-newStart) [reply javascript/javascript1/week4/voice_assistant.js:176]: Perfect! Thank you for sharing!

**Pull request merged**